src/EventStore.Client/GossipChannelSelector.cs

```diff
@@ -0,0 +1,213 @@
+			ClusterMessages.VNodeState.Clone,
+			ClusterMessages.VNodeState.DiscoverLeader
+		};
+
+		private readonly Dictionary<EndPoint, ChannelBase> _grpcChannelsCache;
+		private readonly Dictionary<EndPoint, ServerCapabilities> _serverCapabilitiesCache;
+		private readonly System.Threading.Channels.Channel<ChannelInfo> _channelInfoChannel;
+		private readonly System.Threading.Channels.Channel<ControlMessage> _controlChannel;
+		private readonly ILogger<GossipChannelSelector> _log;
+		private int _disposed;
+
+		public GossipChannelSelector(EventStoreClientSettings settings, IGossipClient gossipClient,
+			IServerCapabilitiesClient serverCapabilities, CancellationToken cancellationToken = default) {
+			_settings = settings;
+			_serverCapabilities = serverCapabilities;
+			_cancellationToken = cancellationToken;
+			_grpcChannelsCache = new Dictionary<EndPoint, ChannelBase>();
+			_serverCapabilitiesCache = new Dictionary<EndPoint, ServerCapabilities>();
+			_channelInfoChannel = System.Threading.Channels.Channel.CreateUnbounded<ChannelInfo>();
+			_controlChannel = System.Threading.Channels.Channel.CreateUnbounded<ControlMessage>();
+			_log = settings.LoggerFactory?.CreateLogger<GossipChannelSelector>() ??
+			       new NullLogger<GossipChannelSelector>();
+
+			foreach (var endPoint in settings.ConnectivitySettings.GossipSeeds) {
+				_grpcChannelsCache.Add(endPoint, ChannelFactory.CreateChannel(settings, endPoint));
+			}
+
+			_controlChannel.Writer.TryWrite(ControlMessage.Discover.Instance);
+
+			_ = Task.Run(async () => {
+				var comparer = settings.ConnectivitySettings.NodePreference switch {
+					NodePreference.Leader => NodePreferenceComparers.Leader,
+					NodePreference.Follower => NodePreferenceComparers.Follower,
+					NodePreference.ReadOnlyReplica => NodePreferenceComparers.ReadOnlyReplica,
+					_ => NodePreferenceComparers.Random
+				};
+
+				Loop:
+				while (await _controlChannel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false)) {
+					var message = await _controlChannel.Reader.ReadAsync(cancellationToken).ConfigureAwait(false);
+
+					if (message is ControlMessage.SetEndPoint setEndPoint) {
+						await UseEndPoint(setEndPoint.EndPoint).ConfigureAwait(false);
+					} else {
+						for (var attempt = 1; attempt <= settings.ConnectivitySettings.MaxDiscoverAttempts; attempt++) {
+							var (success, endPoint, nodes) = await TryDiscover(attempt).ConfigureAwait(false);
+
+							if (!success) {
+								continue;
+							}
+
+							await UseEndPoint(endPoint!).ConfigureAwait(false);
+
+							foreach (var endpoint in _grpcChannelsCache.Keys.ToArray()) {
+								if (nodes!.Any(node =>
+									    node.EndPoint.Host == endpoint.GetHost() &&
+									    node.EndPoint.Port == endpoint.GetPort())) {
+									continue;
+								}
+
+								if (!_grpcChannelsCache.TryGetValue(endpoint, out var channel)) {
+									continue;
+								}
+
+								_grpcChannelsCache.Remove(endpoint);
```

Comment:
i think this is removing all the cached channels except those that correpond to nodes that are both:
1. in the current gossip and
2. in an 'allowed' state

but, if a node is removed because its in a notallowed state like PreReplica, im not sure that it'll get added back to the cache again unless it becomes leader.

im slightly suspicious because we use the cache to determine who to query for gossip. im not sure if we perhaps want to not clean the cache based on node state, or perhaps replenish the cache from the settings, or perhaps use the settings instead of the cache to determine who to query for gossip (?)